sklearn/feature_extraction/image.py

```diff
@@ -125,19 +129,34 @@ def _to_graph(
     diag_idx = np.arange(n_voxels)
     i_idx = np.hstack((edges[0], edges[1]))
     j_idx = np.hstack((edges[1], edges[0]))
+
+    diag_array = diag._array if hasattr(diag, '_array') else diag
+    xp, _ = get_namespace(diag)
     graph = sparse.coo_matrix(
         (
-            np.hstack((weights, weights, diag)),
-            (np.hstack((i_idx, diag_idx)), np.hstack((j_idx, diag_idx))),
+            xp.asarray(np.hstack((weights, weights, diag_array))),
+            (xp.asarray(np.hstack((i_idx, diag_idx))), xp.asarray(np.hstack((j_idx, diag_idx)))),
         ),
         (n_voxels, n_voxels),
         dtype=dtype,
     )
+    if xp.__name__ == 'cupy.array_api':
+        from cupyx.scipy import sparse as cupy_sparse
+        # TODO: should create directly instead of converting
+        graph = cupy_sparse.coo_matrix(graph)
     if return_as is np.ndarray:
         return graph.toarray()
     return return_as(graph)
 
 
+def _expand_dim_to_atleast3d(xp, img):
```

Comment:
Another function that needs to either be wrapped or have a helper function. Depending on how we decide on https://github.com/Quansight-Labs/amd-project/issues/17